### DISCUSSION TITLE
SPI limiting esp32s3 to a baud rate of 800,000

### DISCUSSION THREAD

**Discussion by MATTYGILO** [micropython/micropython#11328] [ESP32]
I am using this code on an esp32s3

```
baudrate = 40_000_000
spisd = SoftSPI(baudrate= baudrate, sck=sck.pin, mosi=mosi.pin, miso=miso.pin)
print("Using SPI: {}".format(spisd))
```

Outputs:

> Using SPI: SoftSPI(baudrate=800000, polarity=0, phase=0, sck=12, mosi=11, miso=13)

As you can see 800_000 != 40_000_000

When I use a value < 800_000. For example 200_000.
> SoftSPI(baudrate=166666, polarity=0, phase=0, sck=12, mosi=11, miso=13)
166_666 is close to the value of 200_000. 

## Conclusion: Micropython is limiting the ESP32S3 to a baud rate of 800_000

All help is much appreciated, thanks in advance

**Comment by robert-hh** [comment]
SoftSPI is, what it says: SPI implemented in software. Given that the CPU runs at 160MHz, you cannot expect a 40MHz baudrate. And generally for SPI the baud rates cannot be precise. Did you try the same with hardware SPI? The SPI hardware is clocked from a 160MHz clock, AFAIK. There, 40Mhz or even 80Mhz are accepted as input value.
Edit: Tested with a scope: SCK is 80 MHz or 40MHz, when set so.

**Comment by MATTYGILO** [comment]
@robert-hh Yes I have tried normal SPI also, makes no difference

And micropython also just says:

    Warning: SPI(-1, ...) is deprecated, use SoftSPI(...) instead

**Comment by robert-hh** [comment]
> Warning: SPI(-1, ...) is deprecated,

SPI(-1. ...) was used to denote SoftSPI. So with SPI(-1, ...) you still got SoftSPI. For hard SPI, use e.g. SPI(0, ....) or SPI(1, .....). And I checked with the oscilloscope, that with hard SPI SCK is indeed 40Mhz, if selected.

**Comment by MATTYGILO** [comment]
@robert-hh  What ID should I use in micropython for SPI pins 11-13
![ESP32-S3_DevKitC-1_pinlayout_v1 1](https://user-images.githubusercontent.com/34808802/234084750-496fb0b2-c9d4-41f9-9395-1df626961054.jpg)

**Comment by robert-hh** [comment]
That's SPI(0). If you write:
```
from machine import SPI
spi=SPI(0, baudrate=40_000_000)
print(spi)
```
it will show you which pins are used.

**Comment by MATTYGILO** [comment]
@robert-hh So I tried the suggestion but get this error?

    ValueError: SPI(0) doesn't exist

**Reply by MATTYGILO** [reply]
If it help's I am using GENERIC_S3_SPIRAM

**Reply by MATTYGILO** [reply]
Seems that those configuration for the pins are not currently available in micropython for the ESP32S3

https://github.com/micropython/micropython/blob/9e6885ad820a9f42bf93df56eb3f3be0c8639622/ports/esp32/machine_hw_spi.c#L73
https://github.com/micropython/micropython/blob/9e6885ad820a9f42bf93df56eb3f3be0c8639622/ports/esp32/boards/GENERIC_S3/mpconfigboard.h

Chat GPT: 
```
Based on the pin assignments you provided in your previous message (Pin 13 for MISO, Pin 11 for MOSI, Pin 12 for SCLK, and Pin 10 for CS), it seems that you are using a non-standard pin mapping, which may not align with the default SPI bus numbering in MicroPython for ESP32S3.
```

**Reply by robert-hh** [reply]
I see, I used SPI(1) on a Lolin S3 boards.

From the answer of ChatGPT It sounds as if it went into politics: many words but no answer,

**Reply by MATTYGILO** [reply]
I've added:

```
#define MICROPY_HW_SPI1_SCK (12)
#define MICROPY_HW_SPI1_MOSI (11)
#define MICROPY_HW_SPI1_MISO (13)
```
To my mpconfigboard.h. Going to see if it makes any difference

**Reply by MATTYGILO** [reply]
I tried SPI2 version, but it failed also

**Reply by beyonlo** [reply]
> I tried SPI2 version, but it failed also

@MATTYGILO  What MicroPython version are you using? here the `SPI(1)` and `SPI(2)` works, but `SPI(0)` fail:

```Python

>>> import os
>>> os.uname()
(sysname='esp32', nodename='esp32', release='1.19.1', version='v1.19.1-966-g05bb26010 on 2023-03-13', machine='ESP32S3 module (spiram) with ESP32S3')
>>> from machine import SPI
>>> spi=SPI(0, baudrate=40_000_000)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: SPI(0) doesn't exist
>>> 
>>> spi=SPI(1, baudrate=40_000_000)
>>> print(spi)
SPI(id=1, baudrate=40000000, polarity=0, phase=0, bits=8, firstbit=0, sck=12, mosi=11, miso=13)
>>> spi=SPI(2, baudrate=40_000_000)
>>> print(spi)
SPI(id=2, baudrate=40000000, polarity=0, phase=0, bits=8, firstbit=0, sck=36, mosi=35, miso=37)
>>> spi=SPI(3, baudrate=40_000_000)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: SPI(3) doesn't exist
>>> 

```
